graql/reasoner/atom/binary/RelationAtom.java

```diff
@@ -453,7 +453,9 @@ private boolean isTypeRoleCompatible(Variable typedVar, Type parentType, boolean
 
         if (roleRequirements.isEmpty()) return true;
 
-        Set<Type> parentTypes = parentType.subs().collect(Collectors.toSet());
+        // in some cases, the parent types are not specified by ISA but by ID directly - then we do no have to search subtypes for compatibility
+        Set<Type> parentTypes = parentTypeIsExact ? Sets.newHashSet(parentType) : parentType.subs().collect(Collectors.toSet());
```

Comment:
You added a boolean and pick the `isa` type if id is present. What happens in the following scenario:
```
($x, $y) isa someRelation;" +
$x id instanceIdOfSubTypeThatCanPlayRolesOfX;
$x isa superTypeThatCannotPlayRolesOfX;
```
or the case when the sub-super types are inverted.

Also I've just realised that the MetaType filter at the top might be invalid -> it assumes all meta types can play all the roles which doesn't have to be the case (even for `entity`).

Reply:
solved by passing the exact type instead of a boolea !